common/djangoapps/microsite_configuration/models.py

```diff
@@ -0,0 +1,181 @@
+        - The values field must be validated on save to prevent the platform from crashing
+        badly in the case the string is not able to be loaded as json.
+    """
+    key = models.CharField(max_length=63, db_index=True, unique=True)
+    subdomain = models.CharField(max_length=127, db_index=True, unique=True)
+    values = JSONField(null=False, blank=True, load_kwargs={'object_pairs_hook': collections.OrderedDict})
+
+    def __unicode__(self):
+        return self.key
+
+    def get_orgs(self):
+        """
+        Helper method to return a list of ORGs associated with our particular Microsite
+        """
+        return MicrositeOrgMapping.get_orgs_for_microsite_by_pk(self.id)  # pylint: disable=no-member
+
+    @classmethod
+    def get_microsite_for_domain(cls, domain):
+        """
+        Returns the microsite associated with this domain. Note that we always convert to lowercase, or
+        None if no match
+        """
+
+        # remove any port number from the hostname
+        domain = domain.split(':')[0]
+        microsites = cls.objects.filter(subdomain__iexact=domain)
+
+        return microsites[0] if microsites else None
+
+
+class MicrositeHistory(TimeStampedModel):
+    """
+    This is an archive table for Microsites model, so that we can maintain a history of changes. Note that the
+    key field is no longer unique
+    """
+    key = models.CharField(max_length=63, db_index=True)
+    subdomain = models.CharField(max_length=127, db_index=True)
+    values = JSONField(null=False, blank=True, load_kwargs={'object_pairs_hook': collections.OrderedDict})
+
+    def __unicode__(self):
+        return self.key
+
+    class Meta(object):
+        """ Meta class for this Django model """
+        verbose_name_plural = "Microsite histories"
+
+
+def _make_archive_copy(instance):
+    """
+    Helper method to make a copy of a Microsite into the history table
+    """
+    archive_object = MicrositeHistory(
+        key=instance.key,
+        subdomain=instance.subdomain,
+        values=instance.values,
+    )
+    archive_object.save()
+
+
+@receiver(pre_delete, sender=Microsite)
+def on_microsite_deleted(sender, instance, **kwargs):  # pylint: disable=unused-argument
+    """
+    Archive the exam attempt when the item is about to be deleted
+    Make a clone and populate in the History table
+    """
+    _make_archive_copy(instance)
+
+
+@receiver(pre_save, sender=Microsite)
+def on_microsite_updated(sender, instance, **kwargs):  # pylint: disable=unused-argument
+    """
+    Archive the microsite on an update operation
+    """
+
+    if instance.id:
+        # on an update case, get the original and archive it
+        original = Microsite.objects.get(id=instance.id)
+        _make_archive_copy(original)
+
+
+class MicrositeOrgMapping(models.Model):
+    """
+    Mapping of ORG to which Microsite it belongs
+    """
+
+    org = models.CharField(max_length=63, db_index=True, unique=True)
```

Comment:
I'd prefer we name these variables 'organization' versus 'org' in the code, but that's more of a nit vs. the class names